<a href="https://colab.research.google.com/github/EmanHrustemovic/FlyRank-AI-Intership-ML-Track-/blob/main/work/notebooks/w03_data_contract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-04 — Search Intelligence Data Contract

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [29]:
from huggingface_hub import HfApi
from google.colab import userdata

hf_token = userdata.get('eman')

api = HfApi(token=hf_token)
files = api.list_repo_files("FlyRank/internship-warehouse", repo_type="dataset")

for f in files:
    print(f)

.gitattributes
README.md
dim_clients.parquet
dim_content.parquet
fact_content_daily_performance/month=2025-01/data_0.parquet
fact_content_daily_performance/month=2025-02/data_0.parquet
fact_content_daily_performance/month=2025-03/data_0.parquet
fact_content_daily_performance/month=2025-04/data_0.parquet
fact_content_daily_performance/month=2025-05/data_0.parquet
fact_content_daily_performance/month=2025-06/data_0.parquet
fact_content_daily_performance/month=2025-07/data_0.parquet
fact_content_daily_performance/month=2025-08/data_0.parquet
fact_content_daily_performance/month=2025-09/data_0.parquet
fact_content_daily_performance/month=2025-10/data_0.parquet
fact_content_daily_performance/month=2025-11/data_0.parquet
fact_content_daily_performance/month=2025-12/data_0.parquet
fact_content_daily_performance/month=2026-01/data_0.parquet
fact_content_daily_performance/month=2026-02/data_0.parquet
fact_content_daily_performance/month=2026-03/data_0.parquet
fact_content_daily_performance/mont

In [30]:
import duckdb
from google.colab import userdata

con = duckdb.connect()
hf_token = userdata.get('eman')  # matches your actual secret name
con.execute(f"CREATE SECRET (TYPE huggingface, TOKEN '{hf_token}')")

rel = "hf://datasets/FlyRank/internship-warehouse"

rel = "hf://datasets/FlyRank/internship-warehouse"

print("=== fact_content_daily_performance (sample - mechanics test only) ===")
display(con.sql(f"DESCRIBE SELECT * FROM read_parquet('{rel}/fact_content_daily_performance_sample.parquet') LIMIT 1"))

print("=== dim_content ===")
display(con.sql(f"DESCRIBE SELECT * FROM read_parquet('{rel}/dim_content.parquet') LIMIT 1"))

print("=== fact_content_query_90d ===")
display(con.sql(f"DESCRIBE SELECT * FROM read_parquet('{rel}/fact_content_query_90d.parquet') LIMIT 1"))

print("=== dim_clients ===")
display(con.sql(f"DESCRIBE SELECT * FROM read_parquet('{rel}/dim_clients.parquet') LIMIT 1"))

print("=== fact_content_daily_performance, month=2026-03 (your working month) ===")
display(con.sql(f"DESCRIBE SELECT * FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet') LIMIT 1"))

=== fact_content_daily_performance (sample - mechanics test only) ===


┌────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│    column_name     │ column_type │  null   │   key   │ default │  extra  │
│      varchar       │   varchar   │ varchar │ varchar │ varchar │ varchar │
├────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ report_date        │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ client_hash_id     │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ content_hash_id    │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ client_has_gsc     │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ client_has_ga4     │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ gsc_data_available │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ ga4_data_available │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ gsc_impressions    │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ gsc_clicks         │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │

=== dim_content ===


┌────────────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│        column_name         │ column_type │  null   │   key   │ default │  extra  │
│          varchar           │   varchar   │ varchar │ varchar │ varchar │ varchar │
├────────────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ client_hash_id             │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ content_hash_id            │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ keyword_hash_id            │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ url_hash_id                │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ keyword_char_count         │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ keyword_token_count        │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ url_char_count             │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ content_created_date       │ DATE        │ YES     │ NULL    │ 

=== fact_content_query_90d ===


┌───────────────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│          column_name          │ column_type │  null   │   key   │ default │  extra  │
│            varchar            │   varchar   │ varchar │ varchar │ varchar │ varchar │
├───────────────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ client_hash_id                │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ content_hash_id               │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ query_hash_id                 │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ query_char_count              │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ query_token_count             │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ window_start                  │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ window_end                    │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ impressions_90d               

=== dim_clients ===


┌─────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│     column_name     │ column_type │  null   │   key   │ default │  extra  │
│       varchar       │   varchar   │ varchar │ varchar │ varchar │ varchar │
├─────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ client_hash_id      │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ is_active           │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ has_gsc_access      │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ has_ga4_access      │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ access_profile      │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ client_created_date │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ client_updated_date │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ gsc_data_start      │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ ga4_data_start      │ DATE        │ YES     │ NULL    │ NULL  

=== fact_content_daily_performance, month=2026-03 (your working month) ===


┌────────────────────┬─────────────┬─────────┬─────────┬─────────┬─────────┐
│    column_name     │ column_type │  null   │   key   │ default │  extra  │
│      varchar       │   varchar   │ varchar │ varchar │ varchar │ varchar │
├────────────────────┼─────────────┼─────────┼─────────┼─────────┼─────────┤
│ report_date        │ DATE        │ YES     │ NULL    │ NULL    │ NULL    │
│ client_hash_id     │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ content_hash_id    │ VARCHAR     │ YES     │ NULL    │ NULL    │ NULL    │
│ client_has_gsc     │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ client_has_ga4     │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ gsc_data_available │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ ga4_data_available │ BOOLEAN     │ YES     │ NULL    │ NULL    │ NULL    │
│ gsc_impressions    │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │
│ gsc_clicks         │ BIGINT      │ YES     │ NULL    │ NULL    │ NULL    │

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

In [31]:
print("Explained below :")

Explained below :


"One row represents one pseudonymized content item, aggregated over a time window within the mid-panel month (March 2026), with a derived trend signal (rising/falling) used for the is_declining_label.

Time window: daily performance data is aggregated from fact_content_daily_performance for month = 2026-03. I compare the earlier portion of the window (e.g., first 15 days) with the later portion (last 15 days) to derive trend_direction."

## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

In [32]:
print("Explained below:")

Explained below:


Buckets / Fields (preliminary – to be confirmed):

Features:

average_position (early period)

average_position (late period)

total_impressions

total_clicks

CTR

days_since_last_publish_or_update (staleness)

distinct_queries (from fact_content_query_90d)

Label:

is_declining_label = 1 if trend_direction == "down" (derived by comparing early vs. late window periods)

Context:

content_id (pseudonymized)

client_id

content category / page type (if available)

Excluded:

The last ~15 days within the sealed test month (June) – kept aside for validation, not used for training.

Raw query-level rows – too granular for page-level classification; preserved for later validation.

## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

**Query** ***1*** **—** **Grain** **check**


In [33]:
q1 = con.sql(f"""
    SELECT
        COUNT(*) AS total_rows,
        COUNT(DISTINCT (client_hash_id, content_hash_id, report_date)) AS distinct_grain_keys
    FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet')
""")
q1
# If total_rows == distinct_grain_keys, the daily grain claim holds.

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌────────────┬─────────────────────┐
│ total_rows │ distinct_grain_keys │
│   int64    │        int64        │
├────────────┼─────────────────────┤
│    9841378 │             9841378 │
└────────────┴─────────────────────┘

Query 2 — Row count + date span:


In [34]:
q2 = con.sql(f"""
    SELECT
        COUNT(*) AS row_count,
        MIN(report_date) AS min_date,
        MAX(report_date) AS max_date,
        COUNT(DISTINCT content_hash_id) AS distinct_pages
    FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet')
""")
q2

┌───────────┬────────────┬────────────┬────────────────┐
│ row_count │  min_date  │  max_date  │ distinct_pages │
│   int64   │    date    │    date    │     int64      │
├───────────┼────────────┼────────────┼────────────────┤
│   9841378 │ 2026-03-01 │ 2026-03-31 │         331437 │
└───────────┴────────────┴────────────┴────────────────┘

Query 3 — Availability check (IS TRUE):


In [35]:
q3 = con.sql(f"""
    SELECT
        COUNT(*) AS total_rows,
        COUNT(*) FILTER (WHERE gsc_data_available IS TRUE) AS rows_with_gsc,
        COUNT(*) FILTER (WHERE gsc_data_available IS TRUE) * 1.0 / COUNT(*) AS gsc_availability_rate
    FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet')
""")
q3

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌────────────┬───────────────┬───────────────────────┐
│ total_rows │ rows_with_gsc │ gsc_availability_rate │
│   int64    │     int64     │        double         │
├────────────┼───────────────┼───────────────────────┤
│    9841378 │       3611061 │   0.36692635929643186 │
└────────────┴───────────────┴───────────────────────┘

In [36]:
page_level = con.sql(f"""
    WITH daily AS (
        SELECT * FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet')
        WHERE gsc_data_available IS TRUE
    ),
    early AS (
        SELECT content_hash_id,
            SUM(gsc_impressions) AS impr_early,
            SUM(gsc_clicks) AS clicks_early,
            SUM(gsc_sum_position) AS sum_pos_early
        FROM daily WHERE report_date <= DATE '2026-03-15'
        GROUP BY content_hash_id
    ),
    late AS (
        SELECT content_hash_id,
            SUM(gsc_impressions) AS impr_late,
            SUM(gsc_clicks) AS clicks_late,
            SUM(gsc_sum_position) AS sum_pos_late
        FROM daily WHERE report_date > DATE '2026-03-15'
        GROUP BY content_hash_id
    )
    SELECT
        e.content_hash_id,
        e.impr_early, e.clicks_early,
        (e.sum_pos_early * 1.0 / NULLIF(e.impr_early,0)) AS avg_position_early,
        l.impr_late, l.clicks_late,
        (l.sum_pos_late * 1.0 / NULLIF(l.impr_late,0)) AS avg_position_late,
        CASE
            WHEN (l.sum_pos_late * 1.0 / NULLIF(l.impr_late,0))
               > (e.sum_pos_early * 1.0 / NULLIF(e.impr_early,0))
            THEN 'down'   -- higher avg position number = worse ranking = declining
            ELSE 'up'
        END AS trend_direction
    FROM early e
    JOIN late l ON e.content_hash_id = l.content_hash_id
    WHERE e.impr_early > 0 AND l.impr_late > 0
""")
page_level

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

┌──────────────────────────┬────────────┬──────────────┬────────────────────┬───────────┬─────────────┬────────────────────┬─────────────────┐
│     content_hash_id      │ impr_early │ clicks_early │ avg_position_early │ impr_late │ clicks_late │ avg_position_late  │ trend_direction │
│         varchar          │   int128   │    int128    │       double       │  int128   │   int128    │       double       │     varchar     │
├──────────────────────────┼────────────┼──────────────┼────────────────────┼───────────┼─────────────┼────────────────────┼─────────────────┤
│ content_2e6360ad20fd7107 │        219 │            1 │  4.360730593607306 │       680 │           0 │  6.297058823529412 │ down            │
│ content_ac8663da7484669a │         20 │            0 │               4.15 │        14 │           0 │                8.5 │ down            │
│ content_d49a012dcb924e31 │        246 │            0 │  4.808943089430894 │        83 │           0 │  6.108433734939759 │ down            │

"""
Q1 (grain): total_rows = distinct_grain_keys = 9,841,378 → confirms one row =
(client, content, date). Grain claim holds.

Q2 (row count + date span): 9,841,378 rows, spanning 2026-03-01 to 2026-03-31,
across 331,437 distinct pages.

Q3 (availability, IS TRUE): only 36.7% of rows (3,611,061 / 9,841,378) have
gsc_data_available = TRUE. This is a major exclusion driver — see Section 4.

"""

**5 features**

In [37]:
features = con.sql(f"""
    WITH daily AS (
        SELECT * FROM read_parquet('{rel}/fact_content_daily_performance/month=2026-03/data_0.parquet')
        WHERE gsc_data_available IS TRUE
    ),
    early AS (
        SELECT content_hash_id,
            SUM(gsc_impressions) AS impr_early,
            SUM(gsc_clicks) AS clicks_early,
            SUM(gsc_sum_position) AS sum_pos_early
        FROM daily WHERE report_date <= DATE '2026-03-15'
        GROUP BY content_hash_id
    ),
    late AS (
        SELECT content_hash_id,
            SUM(gsc_impressions) AS impr_late,
            SUM(gsc_clicks) AS clicks_late,
            SUM(gsc_sum_position) AS sum_pos_late
        FROM daily WHERE report_date > DATE '2026-03-15'
        GROUP BY content_hash_id
    ),
    labeled AS (
        SELECT
            e.content_hash_id,
            e.impr_early,
            e.clicks_early,
            (e.sum_pos_early * 1.0 / NULLIF(e.impr_early,0)) AS avg_position_early,
            (e.clicks_early * 1.0 / NULLIF(e.impr_early,0)) AS ctr_early,
            (l.sum_pos_late * 1.0 / NULLIF(l.impr_late,0)) AS avg_position_late,  -- NEW: needed for the leak demo
            CASE
                WHEN (l.sum_pos_late * 1.0 / NULLIF(l.impr_late,0))
                   > (e.sum_pos_early * 1.0 / NULLIF(e.impr_early,0))
                THEN 1 ELSE 0
            END AS is_declining_label
        FROM early e
        JOIN late l ON e.content_hash_id = l.content_hash_id
        WHERE e.impr_early > 0 AND l.impr_late > 0
    )
    SELECT
        lb.*,
        dc.word_count,
        dc.content_updated_date,
        DATE '2026-03-15' - dc.content_updated_date AS days_since_update
    FROM labeled lb
    LEFT JOIN read_parquet('{rel}/dim_content.parquet') dc
        ON lb.content_hash_id = dc.content_hash_id
""")
features

┌──────────────────────────┬────────────┬──────────────┬────────────────────┬───────────────────────┬────────────────────┬────────────────────┬────────────┬──────────────────────┬───────────────────┐
│     content_hash_id      │ impr_early │ clicks_early │ avg_position_early │       ctr_early       │ avg_position_late  │ is_declining_label │ word_count │ content_updated_date │ days_since_update │
│         varchar          │   int128   │    int128    │       double       │        double         │       double       │       int32        │   int64    │         date         │       int64       │
├──────────────────────────┼────────────┼──────────────┼────────────────────┼───────────────────────┼────────────────────┼────────────────────┼────────────┼──────────────────────┼───────────────────┤
│ content_04c67f3541177192 │        119 │            1 │ 12.319327731092438 │  0.008403361344537815 │ 15.533018867924529 │                  1 │       3168 │ 2026-02-25           │                18 │


# **Trap — leakage demonstration**


In [38]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_auc_score

df = features.df().dropna()

# Honest model: only early-window features
X_honest = df[['impr_early', 'avg_position_early', 'ctr_early', 'word_count', 'days_since_update']]
y = df['is_declining_label']

X_train, X_test, y_train, y_test = train_test_split(X_honest, y, test_size=0.3, random_state=42)
model = LogisticRegression(max_iter=1000).fit(X_train, y_train)
honest_auc = roc_auc_score(y_test, model.predict_proba(X_test)[:,1])
print("Honest AUC (early-window features only):", honest_auc)

# LEAKY version: add the actual late-window position — this is literally
# what the label was derived from
X_leaky = df[['impr_early', 'avg_position_early', 'ctr_early', 'word_count',
              'days_since_update', 'avg_position_late']]
X_train, X_test, y_train, y_test = train_test_split(X_leaky, y, test_size=0.3, random_state=42)
model_leaky = LogisticRegression(max_iter=1000).fit(X_train, y_train)
leaky_auc = roc_auc_score(y_test, model_leaky.predict_proba(X_test)[:,1])
print("Leaky AUC (with label-derived column):", leaky_auc)

print("\nLesson: leaky AUC jumps toward perfect because avg_position_late is the "
      "exact quantity used to construct is_declining_label. This is a textbook "
      "leakage case — the feature is not knowable at the decision moment (2026-03-15), "
      "it comes from the outcome window itself. Deleting it and keeping the honest "
      "AUC (~0.68) is the correct move.")

FloatProgress(value=0.0, layout=Layout(width='auto'), style=ProgressStyle(bar_color='black'))

Honest AUC (early-window features only): 0.6775136401578024
Leaky AUC (with label-derived column): 0.9999933515461121

Lesson: leaky AUC jumps toward perfect because avg_position_late is the exact quantity used to construct is_declining_label. This is a textbook leakage case — the feature is not knowable at the decision moment (2026-03-15), it comes from the outcome window itself. Deleting it and keeping the honest AUC (~0.68) is the correct move.


## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

This dataset cannot tell me:

1. Unbalanced panel history: per-client history depth differs (dim_clients.
   gsc_data_start / ga4_data_start), so early months are not equally
   represented across all clients — some pages simply have no data before
   a certain date.

2. GSC-only availability gap: only 36.7% of daily rows in month=2026-03
   have gsc_data_available = TRUE (Q3). The remaining ~63% are excluded
   from this analysis, meaning the page-level trend signal is built on a
   minority, non-random subset of daily observations — pages with sparse
   or missing GSC data are systematically underrepresented.

3. Window overlap / short baseline: the early (Mar 1–15) vs late (Mar 16–31)
   split gives each page only ~15 days per side. This is a short baseline
   for a "trend" — it can't distinguish a real ranking decline from normal
   day-to-day search volatility.

4. No dim_clients access: I received a 403 error when querying dim_clients.
   parquet, so client-level context (e.g. industry, GSC/GA4 onboarding
   dates) could not be joined in for this notebook.

5. rare_query_count and anonymized_impressions_share in fact_content_query_90d
   mean some query-level detail is aggregated/masked by Google itself —
   the true long-tail query behavior behind a page's traffic isn't fully
   observable.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.